sklearn/utils/__init__.py

```diff
@@ -180,111 +182,56 @@ def axis0_safe_slice(X, mask, len_mask):
         key_array = np.asarray(key)
         if np.issubdtype(key_array.dtype, np.bool_):
             key = key_array
     return array[key] if axis == 0 else array[:, key]
 
 
-def _safe_indexing_row(X, indices):
-    """Return items or rows from X using indices.
-
-    Allows simple indexing of lists or arrays.
-
-    Parameters
-    ----------
-    X : array-like, sparse-matrix, list, pandas.DataFrame, pandas.Series
-        Data from which to sample rows or items.
-    indices : array-like of int
-        Indices according to which X will be subsampled.
-
-    Returns
-    -------
-    subset
-        Subset of X on first axis.
-
-    Notes
-    -----
-    CSR, CSC, and LIL sparse matrices are supported. COO sparse matrices are
-    not supported.
-    """
-    if hasattr(X, "iloc"):
-        # Work-around for indexing with read-only indices in pandas
-        indices = np.asarray(indices)
-        indices = indices if indices.flags.writeable else indices.copy()
-        # Pandas Dataframes and Series
-        try:
-            return X.iloc[indices]
-        except ValueError:
-            # Cython typed memoryviews internally used in pandas do not support
-            # readonly buffers.
-            warnings.warn("Copying input dataframe for slicing.",
-                          DataConversionWarning)
-            return X.copy().iloc[indices]
-    elif hasattr(X, "shape"):
-        if hasattr(X, 'take') and (hasattr(indices, 'dtype') and
-                                   indices.dtype.kind == 'i'):
-            # This is often substantially faster than X[indices]
-            return X.take(indices, axis=0)
-        else:
-            return _array_indexing(X, indices, axis=0)
+def _pandas_indexing(X, key, axis):
+    """Index a pandas dataframe or a series."""
+    # check whether we should index with loc or iloc
+    if _check_key_type(key, int):
+        by_name = False
+    elif _check_key_type(key, str):
+        by_name = True
+    elif _check_key_type(key, bool):
+        # boolean mask
+        by_name = False
+        if hasattr(X, 'loc'):
+            # pandas boolean masks don't work with iloc, so take loc path
+            by_name = True
     else:
-        return [X[idx] for idx in indices]
+        raise ValueError("No valid specification of the columns. Only a "
+                         "scalar, list or slice of all integers or all "
+                         "strings, or boolean mask is allowed")
+
+    if hasattr(key, 'shape'):
+        # Work-around for indexing with read-only key in pandas
+        # FIXME: solved in pandas 0.25
+        key = np.asarray(key)
+        key = key if key.flags.writeable else key.copy()
+    indexer = 'loc' if by_name else 'iloc'
+    return (getattr(X, indexer)[:, key] if axis else getattr(X, indexer)[key])
+
+
+def _list_indexing(X, key):
+    """Index a Python list."""
+    if not isinstance(key, Iterable) or isinstance(indexable, slice):
```

Comment:
In this case is "not Iterable" == scaler? Can we directly check for key being a scaler?

Reply:
Uhm there is a bug here. The first condition is always True because the second condition calls `indexable` which is not defined. 

> Can we directly check for key being a scaler?

isn't it that a scalar is just not an Iterable? I am not aware of something that could check scalar? Do you have something in mind?

Reply:
Does `np.isscalar` work?

Reply:
And a `isinstance(key, slice)` for the slice part.